File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/library/common/compositions/BaseLibraryCompositionsPresenter.java
package com.github.anrimian.musicplayer.ui.library.common.compositions;

import com.github.anrimian.musicplayer.domain.business.player.MusicPlayerInteractor;
import com.github.anrimian.musicplayer.domain.business.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.business.settings.DisplaySettingsInteractor;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent;
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem;
import com.github.anrimian.musicplayer.domain.models.player.PlayerState;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.utils.TextUtils;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import moxy.MvpPresenter;

import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.dispose;
import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.isInactive;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.asList;

public abstract class BaseLibraryCompositionsPresenter<T extends BaseLibraryCompositionsView>
        extends MvpPresenter<T> {

    private final MusicPlayerInteractor playerInteractor;
    private final PlayListsInteractor playListsInteractor;
    private final DisplaySettingsInteractor displaySettingsInteractor;
    protected final ErrorParser errorParser;
    protected final Scheduler uiScheduler;

    private final CompositeDisposable presenterBatterySafeDisposable = new CompositeDisposable();
    protected final CompositeDisposable presenterDisposable = new CompositeDisposable();

    protected Disposable currentCompositionDisposable;
    private Disposable compositionsDisposable;

    private List<Composition> compositions = new ArrayList<>();
    private final LinkedHashSet<Composition> selectedCompositions = new LinkedHashSet<>();
    private final List<Composition> compositionsForPlayList = new LinkedList<>();
    private final List<Composition> compositionsToDelete = new LinkedList<>();

    @Nullable
    private Composition currentComposition;

    @Nullable
    private String searchText;

    public BaseLibraryCompositionsPresenter(MusicPlayerInteractor playerInteractor,
                                            PlayListsInteractor playListsInteractor,
                                            DisplaySettingsInteractor displaySettingsInteractor,
                                            ErrorParser errorParser,
                                            Scheduler uiScheduler) {
        this.playerInteractor = playerInteractor;
        this.playListsInteractor = playListsInteractor;
        this.displaySettingsInteractor = displaySettingsInteractor;
        this.errorParser = errorParser;
        this.uiScheduler = uiScheduler;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnUiSettings();
        subscribeOnCompositions();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    public void onStart() {
        if (!compositions.isEmpty()) {
            subscribeOnCurrentComposition();
            subscribeOnPlayState();
        }
    }

    public void onStop() {
        presenterBatterySafeDisposable.clear();
    }

    public void onTryAgainLoadCompositionsClicked() {
        subscribeOnCompositions();
    }

    public void onCompositionClicked(int position, Composition composition) {
        if (selectedCompositions.isEmpty()) {
            getViewState().showCompositionActionDialog(composition, position);
        } else {
            if (selectedCompositions.contains(composition)) {
                selectedCompositions.remove(composition);
                getViewState().onCompositionUnselected(composition, position);
            } else {
                selectedCompositions.add(composition);
                getViewState().onCompositionSelected(composition, position);
            }
            getViewState().showSelectionMode(selectedCompositions.size());
        }
    }

    public void onCompositionIconClicked(int position, Composition composition) {
        if (composition.equals(currentComposition)) {
            playerInteractor.playOrPause();
        } else {
            playerInteractor.startPlaying(compositions, position);
            getViewState().showCurrentPlayingComposition(composition);
        }
    }

    public void onCompositionLongClick(int position, Composition composition) {
        selectedCompositions.add(composition);
        getViewState().showSelectionMode(selectedCompositions.size());
        getViewState().onCompositionSelected(composition, position);
    }

    public void onPlayAllButtonClicked() {
        if (selectedCompositions.isEmpty()) {
            playerInteractor.startPlaying(compositions);
        } else {
            playSelectedCompositions();
        }
    }

    public void onDeleteCompositionButtonClicked(Composition composition) {
        compositionsToDelete.clear();
        compositionsToDelete.add(composition);
        getViewState().showConfirmDeleteDialog(compositionsToDelete);
    }

    public void onDeleteSelectedCompositionButtonClicked() {
        compositionsToDelete.clear();
        compositionsToDelete.addAll(selectedCompositions);
        getViewState().showConfirmDeleteDialog(compositionsToDelete);
    }

    public void onDeleteCompositionsDialogConfirmed() {
        deletePreparedCompositions();
    }

    public void onPlayNextCompositionClicked(Composition composition) {
        addCompositionsToPlayNext(asList(composition));
    }

    public void onAddToQueueCompositionClicked(Composition composition) {
        addCompositionsToEnd(asList(composition));
    }

    public void onAddToPlayListButtonClicked(Composition composition) {
        compositionsForPlayList.clear();
        compositionsForPlayList.add(composition);
        getViewState().showSelectPlayListDialog();
    }

    public void onAddSelectedCompositionToPlayListClicked() {
        compositionsForPlayList.clear();
        compositionsForPlayList.addAll(selectedCompositions);
        getViewState().showSelectPlayListDialog();
    }

    public void onPlayListToAddingSelected(PlayList playList) {
        playListsInteractor.addCompositionsToPlayList(compositionsForPlayList, playList)
                .observeOn(uiScheduler)
                .subscribe(() -> onAddingToPlayListCompleted(playList),
                        this::onAddingToPlayListError);
    }

    public void onSelectionModeBackPressed() {
        closeSelectionMode();
    }

    public void onShareSelectedCompositionsClicked() {
        getViewState().shareCompositions(selectedCompositions);
    }

    public void onPlayAllSelectedClicked() {
        playSelectedCompositions();
    }

    public void onSelectAllButtonClicked() {
        selectedCompositions.clear();//reselect previous feature
        selectedCompositions.addAll(compositions);
        getViewState().showSelectionMode(compositions.size());
        getViewState().setItemsSelected(true);
    }

    public void onPlayNextSelectedCompositionsClicked() {
        addCompositionsToPlayNext(new ArrayList<>(selectedCompositions));
        closeSelectionMode();
    }

    public void onAddToQueueSelectedCompositionsClicked() {
        addCompositionsToEnd(new ArrayList<>(selectedCompositions));
        closeSelectionMode();
    }

    public void onPlayActionSelected(int position) {
        playerInteractor.startPlaying(compositions, position);
    }

    public void onSearchTextChanged(String text) {
        if (!TextUtils.equals(searchText, text)) {
            searchText = text;
            subscribeOnCompositions();
        }
    }

    public HashSet<Composition> getSelectedCompositions() {
        return selectedCompositions;
    }

    protected void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showErrorMessage(errorCommand);
    }

    private void onDeleteCompositionError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeleteCompositionError(errorCommand);
    }

    private void onAddingToPlayListError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showAddingToPlayListError(errorCommand);
    }

    private void onAddingToPlayListCompleted(PlayList playList) {
        getViewState().showAddingToPlayListComplete(playList, compositionsForPlayList);
        compositionsForPlayList.clear();

        if (!selectedCompositions.isEmpty()) {
            closeSelectionMode();
        }
    }

    private void addCompositionsToPlayNext(List<Composition> compositions) {
        playerInteractor.addCompositionsToPlayNext(compositions)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    private void addCompositionsToEnd(List<Composition> compositions) {
        playerInteractor.addCompositionsToEnd(compositions)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    private void playSelectedCompositions() {
        playerInteractor.startPlaying(new ArrayList<>(selectedCompositions));
        closeSelectionMode();
    }

    private void closeSelectionMode() {
        selectedCompositions.clear();
        getViewState().showSelectionMode(0);
        getViewState().setItemsSelected(false);
    }

    private void deletePreparedCompositions() {
        playerInteractor.deleteCompositions(compositionsToDelete)
                .observeOn(uiScheduler)
                .subscribe(this::onDeleteCompositionsSuccess, this::onDeleteCompositionError);
    }

    private void onDeleteCompositionsSuccess() {
        getViewState().showDeleteCompositionMessage(compositionsToDelete);
        compositionsToDelete.clear();

        if (!selectedCompositions.isEmpty()) {
            closeSelectionMode();
        }
    }

    protected void subscribeOnCurrentComposition() {
        currentCompositionDisposable = playerInteractor.getCurrentCompositionObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onCurrentCompositionReceived, errorParser::logError);
        presenterBatterySafeDisposable.add(currentCompositionDisposable);
    }

    protected void subscribeOnPlayState() {
        presenterBatterySafeDisposable.add(playerInteractor.getPlayerStateObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onPlayerStateReceived));
    }

    protected void subscribeOnCompositions() {
        if (compositions.isEmpty()) {
            getViewState().showLoading();
        }
        dispose(compositionsDisposable, presenterDisposable);
        compositionsDisposable = getCompositionsObservable(searchText)
                .observeOn(uiScheduler)
                .subscribe(this::onCompositionsReceived, this::onCompositionsReceivingError);
        presenterDisposable.add(compositionsDisposable);
    }

    private void onCompositionsReceivingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showLoadingError(errorCommand);
    }

    private void onCompositionsReceived(List<Composition> compositions) {
        this.compositions = compositions;
        getViewState().updateList(compositions);
        if (compositions.isEmpty()) {
            if (TextUtils.isEmpty(searchText)) {
                getViewState().showEmptyList();
            } else {
                getViewState().showEmptySearchResult();
            }
        } else {
            getViewState().showList();

            if (isInactive(currentCompositionDisposable)) {
                subscribeOnCurrentComposition();
                subscribeOnPlayState();
            }
        }
    }

    private void onPlayerStateReceived(PlayerState state) {
        getViewState().showPlayState(state == PlayerState.PLAY);
    }

    private void onCurrentCompositionReceived(PlayQueueEvent playQueueEvent) {
        PlayQueueItem queueItem = playQueueEvent.getPlayQueueItem();
        if (queueItem != null) {
            currentComposition = queueItem.getComposition();
        } else {
            currentComposition = null;
        }
        getViewState().showCurrentPlayingComposition(currentComposition);
    }

    private void subscribeOnUiSettings() {
        presenterDisposable.add(displaySettingsInteractor.getCoversEnabledObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onUiSettingsReceived, errorParser::logError));
    }

    private void onUiSettingsReceived(boolean isCoversEnabled) {
        getViewState().setDisplayCoversEnabled(isCoversEnabled);
    }

    protected abstract Observable<List<Composition>> getCompositionsObservable(String searchText);

}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/library/common/compositions/BaseLibraryCompositionsPresenter.kt
package com.github.anrimian.musicplayer.ui.library.common.compositions

import com.github.anrimian.filesync.SyncInteractor
import com.github.anrimian.musicplayer.domain.Constants
import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.composition.CurrentComposition
import com.github.anrimian.musicplayer.domain.models.composition.DeletedComposition
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.domain.models.sync.FileKey
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.domain.utils.ListUtils
import com.github.anrimian.musicplayer.domain.utils.TextUtils
import com.github.anrimian.musicplayer.domain.utils.rx.RxUtils
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.library.common.library.BaseLibraryPresenter
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.Disposable
import java.util.LinkedList

abstract class BaseLibraryCompositionsPresenter<C : Composition, V : BaseLibraryCompositionsView<C>>(
    private val displaySettingsInteractor: DisplaySettingsInteractor,
    private val syncInteractor: SyncInteractor<FileKey, *, Long>,
    private val playerInteractor: LibraryPlayerInteractor,
    playListsInteractor: PlayListsInteractor,
    errorParser: ErrorParser,
    uiScheduler: Scheduler,
) : BaseLibraryPresenter<V>(playerInteractor, playListsInteractor, uiScheduler, errorParser) {

    private var currentCompositionDisposable: Disposable? = null
    private var compositionsDisposable: Disposable? = null

    private var compositions: List<C> = ArrayList()
    private val selectedCompositions = LinkedHashSet<C>()
    private val compositionsForPlayList: MutableList<Composition> = LinkedList()
    private val compositionsToDelete: MutableList<Composition> = LinkedList()

    private var currentComposition: Composition? = null

    //we can replace it with subj
    private var searchText: String? = null

    private var lastDeleteAction: Completable? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnUiSettings()
        subscribeOnRepeatMode()
        subscribeOnCompositions()
        subscribeOnCurrentComposition()
        syncInteractor.getFilesSyncStateObservable()
            .unsafeSubscribeOnUi(viewState::showFilesSyncState)
    }

    fun onStop(listPosition: ListPosition) {
        saveListPosition(listPosition)
    }

    fun onTryAgainLoadCompositionsClicked() {
        subscribeOnCompositions()
    }

    fun onCompositionClicked(position: Int, composition: C) {
        if (selectedCompositions.isEmpty()) {
            if (composition == currentComposition) {
                playerInteractor.playOrPause()
            } else {
                startPlaying(compositions, position)
                viewState.showCurrentComposition(CurrentComposition(composition, true))
            }
            return
        }
        if (selectedCompositions.contains(composition)) {
            selectedCompositions.remove(composition)
            viewState.onCompositionUnselected(composition, position)
        } else {
            selectedCompositions.add(composition)
            viewState.onCompositionSelected(composition, position)
        }
        viewState.showSelectionMode(selectedCompositions.size)
    }

    fun onCompositionIconClicked(position: Int, composition: C) {
        if (composition == currentComposition) {
            playerInteractor.playOrPause()
        } else {
            startPlaying(compositions, position)
            viewState.showCurrentComposition(CurrentComposition(composition, true))
        }
    }

    fun onCompositionLongClick(position: Int, composition: C) {
        selectedCompositions.add(composition)
        viewState.showSelectionMode(selectedCompositions.size)
        viewState.onCompositionSelected(composition, position)
    }

    fun onPlayAllButtonClicked() {
        if (selectedCompositions.isEmpty()) {
            startPlaying(compositions)
        } else {
            playSelectedCompositions()
        }
    }

    fun onDeleteCompositionButtonClicked(composition: Composition) {
        compositionsToDelete.clear()
        compositionsToDelete.add(composition)
        viewState.showConfirmDeleteDialog(compositionsToDelete)
    }

    fun onDeleteSelectedCompositionButtonClicked() {
        compositionsToDelete.clear()
        compositionsToDelete.addAll(selectedCompositions)
        viewState.showConfirmDeleteDialog(compositionsToDelete)
    }

    fun onDeleteCompositionsDialogConfirmed() {
        deletePreparedCompositions()
    }

    fun onPlayNextCompositionClicked(position: Int) {
        val composition = compositions.elementAtOrNull(position)
        if (composition != null) {
            onPlayNextCompositionClicked(composition)
        }
    }

    fun onPlayNextCompositionClicked(composition: Composition) {
        addCompositionsToPlayNext(ListUtils.asList(composition))
    }

    fun onAddToQueueCompositionClicked(composition: Composition) {
        addCompositionsToEndOfQueue(ListUtils.asList(composition))
    }

    fun onAddToPlayListButtonClicked(composition: Composition) {
        compositionsForPlayList.clear()
        compositionsForPlayList.add(composition)
        viewState.showSelectPlayListDialog()
    }

    fun onAddSelectedCompositionToPlayListClicked() {
        compositionsForPlayList.clear()
        compositionsForPlayList.addAll(selectedCompositions)
        viewState.showSelectPlayListDialog()
    }

    fun onPlayListToAddingSelected(playList: PlayList) {
        performAddToPlaylist(compositionsForPlayList, playList, ::onAddingToPlayListCompleted)
    }

    fun onSelectionModeBackPressed() {
        closeSelectionMode()
    }

    fun onShareSelectedCompositionsClicked() {
        viewState.shareCompositions(selectedCompositions)
    }

    fun onPlayAllSelectedClicked() {
        playSelectedCompositions()
    }

    fun onSelectAllButtonClicked() {
        selectedCompositions.clear() //reselect previous feature
        selectedCompositions.addAll(compositions)
        viewState.showSelectionMode(compositions.size)
        viewState.setItemsSelected(true)
    }

    fun onPlayNextSelectedCompositionsClicked() {
        addCompositionsToPlayNext(ArrayList(selectedCompositions))
        closeSelectionMode()
    }

    fun onAddToQueueSelectedCompositionsClicked() {
        addCompositionsToEndOfQueue(ArrayList(selectedCompositions))
        closeSelectionMode()
    }

    fun onPlayActionSelected(position: Int) {
        startPlaying(compositions, position)
    }

    fun onSearchTextChanged(text: String?) {
        if (!TextUtils.equals(searchText, text)) {
            searchText = text
            subscribeOnCompositions()
        }
    }

    fun onRetryFailedDeleteActionClicked() {
        if (lastDeleteAction != null) {
            lastDeleteAction!!
                .doFinally { lastDeleteAction = null }
                .justSubscribe(this::onDeleteCompositionError)
        }
    }

    fun onChangeRandomModePressed() {
        playerInteractor.changeRandomMode()
    }

    fun getSelectedCompositions(): HashSet<C> = selectedCompositions

    fun getSearchText() = searchText

    protected fun subscribeOnCompositions() {
        if (compositions.isEmpty()) {
            viewState.showLoading()
        }
        RxUtils.dispose(compositionsDisposable, presenterDisposable)
        compositionsDisposable = getCompositionsObservable(searchText)
            .observeOn(uiScheduler)
            .subscribe(this::onCompositionsReceived, this::onCompositionsReceivingError)
        presenterDisposable.add(compositionsDisposable!!)
    }

    protected fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showErrorMessage(errorCommand)
    }

    private fun onDeleteCompositionError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeleteCompositionError(errorCommand)
    }

    private fun onAddingToPlayListCompleted() {
        compositionsForPlayList.clear()
        if (selectedCompositions.isNotEmpty()) {
            closeSelectionMode()
        }
    }

    private fun playSelectedCompositions() {
        startPlaying(ArrayList(selectedCompositions))
        closeSelectionMode()
    }

    private fun startPlaying(compositions: List<Composition>, firstPosition: Int = Constants.NO_POSITION) {
        playerInteractor.setCompositionsQueueAndPlay(compositions, firstPosition)
            .runOnUi(viewState::showErrorMessage)
    }

    private fun closeSelectionMode() {
        selectedCompositions.clear()
        viewState.showSelectionMode(0)
        viewState.setItemsSelected(false)
    }

    private fun deletePreparedCompositions() {
        lastDeleteAction = playerInteractor.deleteCompositions(compositionsToDelete)
            .observeOn(uiScheduler)
            .doOnSuccess(this::onDeleteCompositionsSuccess)
            .ignoreElement()
        lastDeleteAction!!.justSubscribe(this::onDeleteCompositionError)
    }

    private fun onDeleteCompositionsSuccess(compositions: List<DeletedComposition>) {
        viewState.showDeleteCompositionMessage(compositions)
        compositionsToDelete.clear()
        if (selectedCompositions.isNotEmpty()) {
            closeSelectionMode()
        }
    }

    private fun subscribeOnCurrentComposition() {
        playerInteractor.getCurrentCompositionObservable()
            .runOnUi(this::onCurrentCompositionReceived, viewState::showErrorMessage)
    }

    private fun onCompositionsReceivingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showLoadingError(errorCommand)
    }

    private fun onCompositionsReceived(compositions: List<C>) {
        val firstReceive = this.compositions.isEmpty()

        this.compositions = compositions
        viewState.updateList(compositions)
        if (compositions.isEmpty()) {
            if (TextUtils.isEmpty(searchText)) {
                viewState.showEmptyList()
            } else {
                viewState.showEmptySearchResult()
            }
        } else {
            viewState.showList()
            if (firstReceive) {
                val listPosition = getSavedListPosition()
                if (listPosition != null) {
                    viewState.restoreListPosition(listPosition)
                }
            }

            if (RxUtils.isInactive(currentCompositionDisposable)) {
                subscribeOnCurrentComposition()
            }
        }
    }

    private fun onCurrentCompositionReceived(currentComposition: CurrentComposition) {
        this.currentComposition = currentComposition.composition
        viewState.showCurrentComposition(currentComposition)
    }

    private fun subscribeOnUiSettings() {
        displaySettingsInteractor.getCoversEnabledObservable()
            .subscribeOnUi(viewState::setDisplayCoversEnabled, errorParser::logError)
    }

    private fun subscribeOnRepeatMode() {
        playerInteractor.getRandomPlayingObservable()
            .subscribeOnUi(viewState::showRandomMode, errorParser::logError)
    }

    protected abstract fun getCompositionsObservable(searchText: String?): Observable<List<C>>
    protected abstract fun getSavedListPosition(): ListPosition?
    protected abstract fun saveListPosition(listPosition: ListPosition)

}